In [2]:
import mytools
import pandas as pd

In [3]:
# 读取数据表
数据表1 = mytools.读取SPSS数据文件(R'demo.sav')

In [4]:
数据表1.columns

Index(['序号', '@1、您的性别：', '@2、您的年龄：', '@3、您的婚姻状况：', '@4、您的受教育程度是：',
       '@5、您的政治面貌：', '@7、您的信仰是：', '@8、您的职业是：', '@9、您的月收入大约是多少？', '@10、1您家里有',
       '@10、2口人，是', '@10、3代人，家里有', '@10、4位老人，家里有', '@10、5个小孩，外出打工的人有___人。',
       '@11、您平常是否通过电脑上网？', '@12、您平常是否通过手机上网？', '@13、您平时一周有几天通过电脑上网？',
       '@14、您平时一周有几天通过手机上网？', '@15、1您每天通过手机上网的时间大概为', '@15、2分钟，您每天通过电脑上网的时间为',
       '@15、3分钟，您每个月的手机话费为', '@15、4元，您家的宽带费为___元。',
       '@16、互联网使用行为（问卷17题至26题）—17、您', '@16、18、您认为您在上网时使用搜索引擎（百度',
       '@16、19、您经常会通过互联网（辣椒直播、下厨', '@16、20、您认为您玩手机网络游戏（斗地主、斗',
       '@16、21、您经常会通过互联网观看小说么？', '@16、22、您经常会通过互联网观看视频（快手、',
       '@16、23、您认为您通过即时通讯（微信、QQ）与他', '@16、24、您经常会通过社交网络（微博、陌陌、',
       '@16、25、您认为您在互联网（淘宝、拼多多、京', '@16、26、您经常使用移动支付（支付宝、微信支',
       'Q17_1_我目前的生活接近我理想的状态', 'Q17_2_我的生活条件在各方面都很不错。', 'Q17_3_对我目前的生活，我很满意。',
       'Q17_4_到目前为止，我已经得到了我想要的东西', 'Q17_5_如果我能再活一次我没有什么要改变',
       '@18、请表述您最近一周的情感状况—6.爱', '@18、7.愉悦', '@18、8.自豪', '@18、9.愤怒',
       '@18、10.忧虑', '@18、11.悲哀', '@19、第三部分—1.我能够获得足够

# 明确数据分析目标
1.了解本地区居民的自我感觉（幸福感）
2.研究居民各项数据与自我感觉的相关性
3.居民受教育程度与自尊的相关性
4.大学生，自尊，生活满意度的多变量交互分类表

In [5]:
#换变量名字
数据表1.rename(columns={
    '@1、您的性别：':'性别',
    '@2、您的年龄：': '年龄',
    '@3、您的婚姻状况：': '婚姻状况',
    '@4、您的受教育程度是：': '受教育程度',
    '@9、您的月收入大约是多少？': '月收入',
    '@10、1您家里有': '家庭规模',
    '@10、2口人，是': '代际',
    '@10、5个小孩，外出打工的人有___人。':'外出务工人数',
}, inplace=True)

In [6]:
#平均值
数据表1['年龄'].mean(),数据表1['家庭规模'].mean()

(43.67741935483871, 6.662921348314606)

In [7]:
#把（空）转为None
数据表1['外出务工人数'].replace({'(空)':None}, inplace=True)
数据表1['外出务工人数'].value_counts()

外出务工人数
0    56
1    21
2    10
4     1
Name: count, dtype: int64

In [8]:
数据表1['代际'].replace({'(空)':None}, inplace=True)
数据表1['代际'].value_counts()

代际
3    59
2    18
4     8
1     2
5     1
Name: count, dtype: int64

In [9]:
#填入值
数据表1['年龄'].fillna(43, inplace=True)
数据表1['婚姻状况'].fillna('已婚', inplace=True)
数据表1['月收入'].fillna('三千元以下', inplace=True)
数据表1['家庭规模'].fillna(6, inplace=True)
数据表1['代际'].fillna(3, inplace=True)
数据表1['外出务工人数'].fillna(2,inplace=True)

In [10]:
# 计算均值和标准差
mean = 数据表1['家庭规模'].mean()
std = 数据表1['家庭规模'].std()
condition1 = (数据表1['家庭规模'] < mean - 3 * std) | (数据表1['家庭规模'] > mean + 3 * std)
# 识别异常值
outliers1 = 数据表1[condition1]
print('使用标准差判断的异常值为：',outliers1)

使用标准差判断的异常值为： Empty DataFrame
Columns: [序号, 性别, 年龄, 婚姻状况, 受教育程度, @5、您的政治面貌：, @7、您的信仰是：, @8、您的职业是：, 月收入, 家庭规模, 代际, @10、3代人，家里有, @10、4位老人，家里有, 外出务工人数, @11、您平常是否通过电脑上网？, @12、您平常是否通过手机上网？, @13、您平时一周有几天通过电脑上网？, @14、您平时一周有几天通过手机上网？, @15、1您每天通过手机上网的时间大概为, @15、2分钟，您每天通过电脑上网的时间为, @15、3分钟，您每个月的手机话费为, @15、4元，您家的宽带费为___元。, @16、互联网使用行为（问卷17题至26题）—17、您, @16、18、您认为您在上网时使用搜索引擎（百度, @16、19、您经常会通过互联网（辣椒直播、下厨, @16、20、您认为您玩手机网络游戏（斗地主、斗, @16、21、您经常会通过互联网观看小说么？, @16、22、您经常会通过互联网观看视频（快手、, @16、23、您认为您通过即时通讯（微信、QQ）与他, @16、24、您经常会通过社交网络（微博、陌陌、, @16、25、您认为您在互联网（淘宝、拼多多、京, @16、26、您经常使用移动支付（支付宝、微信支, Q17_1_我目前的生活接近我理想的状态, Q17_2_我的生活条件在各方面都很不错。, Q17_3_对我目前的生活，我很满意。, Q17_4_到目前为止，我已经得到了我想要的东西, Q17_5_如果我能再活一次我没有什么要改变, @18、请表述您最近一周的情感状况—6.爱, @18、7.愉悦, @18、8.自豪, @18、9.愤怒, @18、10.忧虑, @18、11.悲哀, @19、第三部分—1.我能够获得足够有用的媒介信, @19、2.我能够不断更新自己的信息获取技能。, @19、3.我具有较强的信息获取意识。, @19、4.我能够熟练使用软件工具进行信息检索。, @19、5.我能够通过寻找佐证资料判断信息的可信, @19、6.能够通过标题、内容等判断信息的可信度, @19、7.我能够判断当前信息是在陈述事实还是表, @19、8.我能够评估媒介信息内容对他人或社会可, @19、9.我能够通过信息发布机构的权威性判断信, @19、1

In [11]:
mytools.使用标准差判断数值变量异常值(数据表1,'家庭规模')

使用标准差判断的异常值为： Empty DataFrame
Columns: [序号, 性别, 年龄, 婚姻状况, 受教育程度, @5、您的政治面貌：, @7、您的信仰是：, @8、您的职业是：, 月收入, 家庭规模, 代际, @10、3代人，家里有, @10、4位老人，家里有, 外出务工人数, @11、您平常是否通过电脑上网？, @12、您平常是否通过手机上网？, @13、您平时一周有几天通过电脑上网？, @14、您平时一周有几天通过手机上网？, @15、1您每天通过手机上网的时间大概为, @15、2分钟，您每天通过电脑上网的时间为, @15、3分钟，您每个月的手机话费为, @15、4元，您家的宽带费为___元。, @16、互联网使用行为（问卷17题至26题）—17、您, @16、18、您认为您在上网时使用搜索引擎（百度, @16、19、您经常会通过互联网（辣椒直播、下厨, @16、20、您认为您玩手机网络游戏（斗地主、斗, @16、21、您经常会通过互联网观看小说么？, @16、22、您经常会通过互联网观看视频（快手、, @16、23、您认为您通过即时通讯（微信、QQ）与他, @16、24、您经常会通过社交网络（微博、陌陌、, @16、25、您认为您在互联网（淘宝、拼多多、京, @16、26、您经常使用移动支付（支付宝、微信支, Q17_1_我目前的生活接近我理想的状态, Q17_2_我的生活条件在各方面都很不错。, Q17_3_对我目前的生活，我很满意。, Q17_4_到目前为止，我已经得到了我想要的东西, Q17_5_如果我能再活一次我没有什么要改变, @18、请表述您最近一周的情感状况—6.爱, @18、7.愉悦, @18、8.自豪, @18、9.愤怒, @18、10.忧虑, @18、11.悲哀, @19、第三部分—1.我能够获得足够有用的媒介信, @19、2.我能够不断更新自己的信息获取技能。, @19、3.我具有较强的信息获取意识。, @19、4.我能够熟练使用软件工具进行信息检索。, @19、5.我能够通过寻找佐证资料判断信息的可信, @19、6.能够通过标题、内容等判断信息的可信度, @19、7.我能够判断当前信息是在陈述事实还是表, @19、8.我能够评估媒介信息内容对他人或社会可, @19、9.我能够通过信息发布机构的权威性判断信, @19、1

In [12]:
#数据类型
数据表1['外出务工人数'].dtypes

dtype('O')

In [13]:
#数据类型
数据表1['家庭规模'].dtypes

dtype('float64')

In [14]:
#数据类型转化
数据表1.astype({"外出务工人数": "int"}).dtypes

序号          float64
性别          float64
年龄          float64
婚姻状况       category
受教育程度      category
             ...   
生活满意度       float64
生活满意度区间    category
工作满意度       float64
工作满意度区间    category
自尊          float64
Length: 89, dtype: object

In [15]:
数据表1['家庭规模'].map({1:1,2:2,3:3,4:4,5:5})

0     NaN
1     NaN
2     NaN
3     NaN
4     NaN
     ... 
89    5.0
90    NaN
91    NaN
92    4.0
93    NaN
Name: 家庭规模, Length: 94, dtype: float64

In [16]:
# 变量类型设置
数据表1['外出务工人数'].dtypes
数据表1.astype({'外出务工人数':"int"}).dtypes

序号          float64
性别          float64
年龄          float64
婚姻状况       category
受教育程度      category
             ...   
生活满意度       float64
生活满意度区间    category
工作满意度       float64
工作满意度区间    category
自尊          float64
Length: 89, dtype: object

In [17]:
# 生成新的变量
数据表1['自我感觉'] = 数据表1['Q17_1_我目前的生活接近我理想的状态']+数据表1['Q17_2_我的生活条件在各方面都很不错。']+数据表1['Q17_3_对我目前的生活，我很满意。']+数据表1['Q17_4_到目前为止，我已经得到了我想要的东西']+数据表1['Q17_5_如果我能再活一次我没有什么要改变']

In [23]:
# 单变量数据分析
mytools.使用plotly绘制类别变量柱状图(数据表1,'自我感觉')

自我感觉良好

In [19]:
# 双变量相关性分析
mytools.类别变量与数值变量统计分析(数据表1, '受教育程度','自我感觉')

各个类别的平均值：
受教育程度
未上过学      24.967742
小学        26.313725
初中        25.285714
高中        19.666667
大学        27.500000
研究生及以上          NaN
Name: 自我感觉, dtype: float64
相关比率：0.07109634265836917
中度相关
F检验的p值：0.1563808146092898
                            OLS Regression Results                            
Dep. Variable:                   自我感觉   R-squared:                       0.071
Model:                            OLS   Adj. R-squared:                  0.029
Method:                 Least Squares   F-statistic:                     1.703
Date:                Thu, 04 Jan 2024   Prob (F-statistic):              0.156
Time:                        18:40:56   Log-Likelihood:                -276.72
No. Observations:                  94   AIC:                             563.4
Df Residuals:                      89   BIC:                             576.2
Df Model:                           4                                         
Covariance Type:            nonrobust                                 

C:\Users\86158\AppData\Roaming\Python\Python38\site-packages\statsmodels\regression\linear_model.py:1965: RuntimeWarning:

divide by zero encountered in scalar divide



根据以上输出结果，我们可以得出以下结论：

1. 受教育程度与自我感觉之间的相关性较低，相关比率为0.071，属于中度相关。

2. 通过F检验的p值为0.156，这意味着受教育程度对自我感觉的影响在统计上不显著。

3. OLS回归结果显示，受教育程度对自我感觉的影响并不显著。各个受教育程度类别的平均值也显示了这一点，因为它们之间的差异并不明显。

需要注意的是，回归结果中的一些系数并不显著，这可能表明存在多重共线性问题或者设计矩阵是奇异的。这可能会影响模型的稳定性和结果的解释。

In [20]:
#双变量
mytools.两个数值变量的统计分析(数据表1,'自我感觉','生活满意度')

决定系数r平方：1.0000
高度相关
p值： 0.0000
拒绝虚无假设


决定系数r平方为1.0000，表示自我感觉和生活满意度之间存在完美的线性关系。同时，p值为0.0000，低于通常的显著性水平（0.05），因此我们可以拒绝虚无假设，即自我感觉和生活满意度之间的相关性是显著的。

In [21]:
mytools.两个无序类别变量的统计分析(数据表1,'受教育程度','自尊')

tau_y系数: 0.0804 极弱相关或无相关
--------  -  -  -  -  -  -  --  --  -  -  -  -  -  -
未上过学  1  1  0  0  0  1   4  22  2  0  0  0  0  0
小学      0  1  2  0  2  1  10  23  3  5  1  2  1  0
初中      0  0  0  1  0  0   1   1  1  0  2  1  0  0
高中      0  0  0  0  0  1   0   2  0  0  0  0  0  0
大学      0  0  0  0  0  0   1   0  0  0  0  0  0  1
--------  -  -  -  -  -  -  --  --  -  -  -  -  -  -
卡方值： 106.98, p值： 0.0000,自由度:52。
拒绝虚无假设


根据卡方检验的结果，'受教育程度'和'自尊'两个变量之间的tau_y系数为0.0804，表明它们之间存在极弱相关或无相关性，很奇怪。具体来说，卡方值为106.98，p值为0.0000，自由度为52。由于p值小于通常的显著性水平（0.05），我们拒绝虚无假设，即'受教育程度'和'自尊'之间的关联在统计上是显著的。
这意味着在这份数据中，受教育程度与自尊之间存在一定程度的关联。这个结果可能需要进一步的研究和分析，以更好地理解这两个变量之间的关系。

In [22]:
#多变量分析
from scipy import stats
# 假设我们关注的是大学生
specific_grade = "大学"

# 筛选出特定年级的数据
df_specific_grade = 数据表1[数据表1['受教育程度'] == specific_grade]

# 创建生活满意度和自我感觉的交互表
cross_tab_specific_grade = pd.crosstab(df_specific_grade['自尊'], df_specific_grade['生活满意度'])

# 进行卡方检验
chi2_grade, p_grade, dof_grade, expected_grade = stats.chi2_contingency(cross_tab_specific_grade)

print(chi2_grade,p_grade)
cross_tab_specific_grade

0.0 1.0


生活满意度  26.0  29.0
自尊               
29.0      0     1
37.0      1     0

根据卡方检验的结果，我们得到了卡方值为0.0和p值为1.0。这表明在特定大学生群体中，自尊和生活满意度之间的关联性不显著。交叉表中的数据展示了自尊为29和37的学生的生活满意度情况，但由于p值高于通常的显著性水平（0.05），我们无法拒绝原假设，即这两个变量之间不存在显著的相关性。

In [24]:
import pandas as pd
from tabulate import tabulate

# 创建一个示例 DataFrame
data = {'Name': ['Alice', 'Bob', 'Charlie'],
        'Age': [25, 30, 35],
        'City': ['New York', 'Los Angeles', 'Chicago']}
df = pd.DataFrame(data)

# 使用 tabulate 打印 DataFrame
print(tabulate(df, headers='keys', tablefmt='psql'))

+----+---------+-------+-------------+
|    | Name    |   Age | City        |
|----+---------+-------+-------------|
|  0 | Alice   |    25 | New York    |
|  1 | Bob     |    30 | Los Angeles |
|  2 | Charlie |    35 | Chicago     |
+----+---------+-------+-------------+
